iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -320,14 +322,106 @@ static Value buildFlowWorkgroupInfoOp(OpBuilder &b, unsigned dim) {
+      insertAfterMap;
+  llvm::SetVector<Operation *> allOps(ops.begin(), ops.end());
+  llvm::SetVector<Operation *> leafOps = allOps;
+  // For each operation compute the list of operations in `ops` that use its
+  // results. Also compute the operations that form the leafs of the DAG of
+  // operations in `ops`.
+  for (auto op : ops) {
+    // insertAfterList.emplace_back(ElementTy{op, {}});
+    // ElementTy &element = insertAfterList.back();
+    for (auto operand : op->getOperands()) {
+      auto definingOp = operand.getDefiningOp();
+      if (!definingOp) continue;
+      insertAfterMap[definingOp].push_back(op);
+      if (allOps.count(definingOp)) leafOps.remove(op);
+    }
+  }
+
+  // The leaves are at the head of the ordered list.
+  SmallVector<Operation *, 16> orderedOps = llvm::to_vector<4>(leafOps);
+  orderedOps.reserve(ops.size());
+  llvm::SmallPtrSet<Operation *, 16> processed;
+  processed.insert(leafOps.begin(), leafOps.end());
+
+  // `readyOps` contains the list of operations that have been just added to the
+  // `orderedOps` list. With these marked ready, they might make further
+  // operations in `ops` ready as well.
+  // The complexity of the algorithm is driven by these
+  // - Each operations is added to `readyOps` list at most once, and is removed
+  //   after being processed
+  // - For every operation in `readyOps` every use of its results (within `ops`)
+  //   is looked at once.
+  // - For every use, the operands of the user are processed.
+  // Assuming operands is O(1), i.e. constant order, the complexity is O(sum of
+  // number of uses of each operation). Given that the size of `ops` is at max
+  // O(10), and not O(100), this is assumed to be reasonable.
+  SmallVector<Operation *, 16> readyOps = orderedOps;
+  readyOps.assign(leafOps.begin(), leafOps.end());
+  while (!readyOps.empty()) {
+    SmallVector<Operation *> nextReadyOps;
+    LLVM_DEBUG({
+      llvm::dbgs() << "ReadyOps :\n";
+      for (auto readyOp : readyOps) {
+        llvm::dbgs() << "\t";
+        readyOp->print(llvm::dbgs());
+        llvm::dbgs() << "\n";
+      }
+    });
+    for (auto readyOp : readyOps) {
+      // Check all uses of `readyOp` within `ops`. If all of the operations that
+      // define the operands of the user have been added to `orderedOps`, then
+      // the user is ready to be scheduled.
+      for (auto insertAfterOp : insertAfterMap[readyOp]) {
+        if (processed.count(insertAfterOp)) continue;
+        if (llvm::all_of(insertAfterOp->getOperands(), [&](Value operand) {
+              Operation *operandDefiningOp = operand.getDefiningOp();
+              return !operandDefiningOp || processed.count(operandDefiningOp);
+            })) {
+          nextReadyOps.push_back(insertAfterOp);
+          orderedOps.push_back(insertAfterOp);
+          processed.insert(insertAfterOp);
+        }
+      }
+    }
+    std::swap(nextReadyOps, readyOps);
+  }
+
+  LLVM_DEBUG({
```

Comment:
this is a lot of debug spew (here and above) - is this enabled with -debug by default?

Reply:
I guess so. Is there a way to hide it even further. I can remove it if it is too much.

Reply:
Looks like DEBUG_WITH_TYPE is the thing to use.

Reply:
Using DEBUG_WITH_TYPE. That makes the debug spew guarded by DEBUG_TYPE.